### ISSUE TITLE
[HOLD for payment 2021-12-15] Unable to use 1Password option for autofill on login 

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#3855]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel! 
___

## Action Performed:
1. Navigate to staging.expensify.cash
2. Right click on the email field
3. Click on the 1Password option

## Expected Result:
Option to select existing accounts should be displayed

## Actual Result:
Nothing happens when clicking on the 1Password option

## Workaround:
No, unable to use the 1Password extension. 

## Platform:
Where is this issue occurring?

 Web ✔️ 
 iOS
 Android
 Desktop App
 Mobile Web

**Version Number:** 1.0.75-5

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

![image](https://user-images.githubusercontent.com/44479856/124169842-00a2e080-daa7-11eb-8eee-514875ad6179.png)

**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

___
From @mallenexpensify https://expensify.slack.com/archives/C01GTK53T8Q/p1625156109326100?thread_ts=1625155719.325000&cid=C01GTK53T8Q

> Be able to right click on the Phone or Email box and have a menu option that allows for selecting 1Password.  OR, be able to see and click a 1Password icon on the far right of the box.
> Ideally, once 1P is used for entering the email address, the Password and Two Factor Code on the following page will be autofilled as well.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @flodnv (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by anthony-hull** [comment]
This is because the password forms on this page don't meet best practices for user password forms:
https://web.dev/sign-in-form-best-practices/
I believe that it won't detect the form because of this.
When I add the password field after the username lastpass detects the inputs and adds the autofill features. So I think it's a detection issue.
It would be good to know is this happens on Android and firefox as they use heuristics for detection.
I'll test 1Password later to see if it behaves the same.

Can I ask if you've experienced 1Password working with split forms before? Once we follow best practices there is still a chance they don't implement support for them. An example of split forms is amazon.com etc. 


Secondly there is another issue that will appear after this one is fix: I don't think it would be able to chose the right password for you when it does trigger as the login is split across two pages. The password manager doesn't have the context it needs to match a user:pass combo.

I saw another symptom of the same issue before when setting a password, it tries to save the password without a username because it doesn't know what the username is in that context.

[If you've implemented an "email first" sign-in flow](https://www.chromium.org/developers/design-documents/form-styles-that-chromium-understands#:~:text=If%20you%27ve%20implemented%20an%20%22email%20first%22%20sign-in%20flow%20that%20separates%20the%20username%20and%20password%20into%20two%20separate%20forms%2C%20include%20a%20form%20field%20containing%20the%20username%20in%20the%20form%20used%20to%20collect%20the%20password.%20You%20can%2C%20of%20course%2C%20hide%20this%20field%20via%20CSS%20if%20that%27s%20appropriate%20for%20your%20layout.)

The saving issue is with chrome password manager and last pass that I've experienced.

For the initial issue I propose further investigating the limitations of detection in the major password managers. Hopefully once the best practices are implemented the split form will work. Otherwise we might need to change the form depending on the limitation of which one we decide to support. which brings me to the next question.
Which managers do you want to support?

Firefox, Chrome, Edge, iOS, Android, Safari, 1Password, LastPass.... the list goes on. I don't see a browserlist query in the project. 

For my second issue:
I propose adding a hidden username form field in the form to let the manager know what password to fill.
I propose also to create another issue to further audit all the password forms in the application to meet best practices e.g. not using placeholders, don't make the user replete themselves (once password managers work correctly).

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kevinksullivan (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by parasharrajat** [comment]
cc: @iwiznia I think this is an important issue and @anthony-hull has a good proposal.

**Comment by iwiznia** [comment]
Thanks for bumping the issue @parasharrajat!
Seems Kevin missed this issue and should've been exported a while ago. Asking someone else to export it and then an engineer will get assigned to review the proposals.

**Comment by parasharrajat** [comment]
It just needs a hidden username input and all password managers should work. If chrome can detect the form then all others will. They follow what chrome does.

**Comment by laurenreidexpensify** [comment]
https://www.upwork.com/jobs/~01b7f83dea2310bcf3 posted to Upwork

**Comment by MelvinBot** [comment]
Triggered auto assignment to @AndrewGable (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by laurenreidexpensify** [comment]
@anthony-hull did you have any proposals to fix this issue?

**Comment by kevinksullivan** [comment]
Sorry I completely missed the export on this, and thanks for stepping in @laurenreidexpensify . Updating my email filters/notifications to make sure this doesn't happen again.

**Comment by laurenreidexpensify** [comment]
No stress @kevinksullivan - pro tip, download the [K2 extension](https://stackoverflow.com/c/expensify/questions/10601) and use it daily and I guarantee you you'll never need a GH notification again to manage your issues :)

**Comment by anthony-hull** [comment]
> It just needs a hidden username input and all password managers should work. If chrome can detect the form then all others will. They follow what chrome does.

as per [chromium's docs](https://www.chromium.org/developers/design-documents/form-styles-that-chromium-understands) I did try adding a hidden username field when I first investigated. It didn't like it! Then suddenly worked once I removed `display: none` I'll have a look at it again 👍

**Comment by MelvinBot** [comment]
Triggered auto assignment to @puneetlath (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by laurenreidexpensify** [comment]
@puneetlath I'm OOO next week - reassigning for someone else from contributor managers to watch over, looks like that's you

**Comment by puneetlath** [comment]
I'm going to be off most of next week as well. I chatted with @kevinksullivan and he's going to take this back.

**Comment by MelvinBot** [comment]
@AndrewGable, @kevinksullivan it looks like no one is assigned to work on this job.
Please double the price or add a comment stating why the job isn't being doubled.

**Comment by kevinksullivan** [comment]
Waiting on proposal from @anthony-hull

**Comment by parasharrajat** [comment]
### Proposal
1. Using `accessibilityRole` we set the containing view element as `Form` in Both SignInPagelayoutWide and SignPagelayoutNarrow.
2. Now we have to use a hidden username input on the PasswordForm page.
```
<View
                        style={{
                            visibility: 'hidden',
                            overflow: 'hidden',
                            width: 0,
                            height: 0,
                        }}
                        accessibilityHidden
                    >
                        <ExpensiTextInput
                            value={this.props.credentials.login}
                            autoCompleteType="username"
                            textContentType="username"
                            nativeID="username"
                            autoCorrect={false}
                        />
                    </View>
```
3. Here credentials can be retrieved using the `credentials: {key: ONYXKEYS.CREDENTIALS}` key.
4. Now we have to use autoCompleteType="username" for the username input on the LoginForm & autoCompleteType="current-password" for password input on the PasswordForm. I will also use the same values with `nativeID` prop.

5. Due to the reason autoCompleteType is different for native and web, I will create a geeneric utility method located at the `TextInputUtils.getPasswordAutocompleteType` . I will also refactor other form related utilities into `TextInputUtils` lib bcz I feel they should be placed in single module.

This should enable the Autofill for password managers.

6. If need arise we have to use `setNativeProps` to set the html attributes which is not supported via react-native. I am skeptical about the `method` attirbute and `autocomplete` on the form but I think they are not necessary AFAIK.

**Comment by anthony-hull** [comment]
Did that work? I tested adding a hidden field and lastpass didn't autofill

**Comment by parasharrajat** [comment]
I think in the video above browser password manager is auto-filling the field.

**Comment by AndrewGable** [comment]
@parasharrajat - Can we confirm @anthony-hull's question? If it does work, we can move forward with your proposal

**Comment by parasharrajat** [comment]
For me it works. I added the video. It is showing that lastpass. @anthony-hull could you please confirm that you were asking the same.

**Comment by AndrewGable** [comment]
👍 - @kevinksullivan can we move forward with @parasharrajat? Thanks!

**Comment by kevinksullivan** [comment]
@parasharrajat can you submit an application so I can hire you?

https://www.upwork.com/ab/applicants/1426160653988712448/job-details

**Comment by parasharrajat** [comment]
Applied and PR submitted for review.

**Comment by kadiealexander** [comment]
Please refer to [this post](https://expensify.slack.com/archives/C01GTK53T8Q/p1633557259458600) for updated information on the `n6 hold`, we've raised the bonus to **$250** for all issues where a **PR is created** before the N6 hold is lifted.

**Comment by mallenexpensify** [comment]
Looks like this one slipped through.  I paid @parasharrajat and closed the contract in Upwork. Closing this issue now, I tested in Chrome and Firefox and using 1Password to log in appears to be working!

**Comment by rushatgabhane** [comment]
Hey @parasharrajat what's the status on this one? 
Asking because it's frustrating when autofill doesn't work.

**Comment by parasharrajat** [comment]
It's in review. I will update the code with new changes this weekend.  So hopefully you will have it next week.

**Comment by parasharrajat** [comment]
Awaiting review...

**Comment by kevinksullivan** [comment]
Looks like this is in requested changes now.

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.18-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/5275


If no regressions arise, payment will be issued on 2021-12-15. :confetti_ball:

**Comment by parasharrajat** [comment]
Thanks, I had to do a lot of refactor to make it work.

**Comment by mallenexpensify** [comment]
@parasharrajat I just tested on Web, Chrome, Staging and - after I used 1Password to add my email address, I wasn't able to click the `return` key to move to the next step (activate the continue button).   I had to click `tab` then `enter` to get to the next step. 

Can this be fixed, updated?  If so... can/should it be done via the current PR or should a new issue be created?

**Comment by parasharrajat** [comment]
Let me check @mallenexpensify

**Comment by mallenexpensify** [comment]
One other thing, which I don't think is directly related to this. Can you test to see, if a user has a 2FA one-time password in 1Password, should that auto-fill on all platforms?  I've had issues with 1P and NewDot on iOS for many months, wondering if others are also having similar issues.  (I think I just got 1P and NewDot iOS working better except that I'm not getting the one-time password added after the password is added)

**Comment by parasharrajat** [comment]
Ok. I got you there @mallenexpensify. I will add that as well but just for the record not part of this issue.

About this https://github.com/Expensify/App/issues/3855#issuecomment-991252825,
It seems to be configured but If could share a video that would be helpful.

**Comment by mallenexpensify** [comment]
> I will add that as well but just for the record not part of this issue.
If the 2FA/one-time-code isn't clearly called out in this issue, does $250 for increased scope seem fair?  Feel free to include in this PR or we can create a new issue.  

For focus not advancing to the `Continue` button after an email address has been entered, he's a vid

https://user-images.githubusercontent.com/22508304/145653682-bc26876f-e1fe-435a-9b22-7f62d7aa6496.mp4

and here's what's showing in logs
![image](https://user-images.githubusercontent.com/22508304/145653585-9f11962b-af5b-4b3b-b728-afaf7b80857d.png)

**Comment by parasharrajat** [comment]
@mallenexpensify pressing enter after email is autofill should take you to the password screen. It's working for me and was working before this issue.

**Comment by parasharrajat** [comment]
Do you want me to take create a PR for https://github.com/Expensify/App/issues/3855#issuecomment-991306682?

**Comment by mallenexpensify** [comment]
> pressing enter after email is autofill should take you to the password screen. It's working for me and was working before this issue.
It's not working for me and I think we should fix it. I'm on staging.new.expensify.com and am using 1Password to autofill.  After I select my email address and hit `enter`, nothing happens, I'm unable to advance to the password screen without clicking the `Continue` box with my cursor.  

> Do you want me to create a PR for [#3855 (comment)](https://github.com/Expensify/App/issues/3855#issuecomment-991306682)?
@AndrewGable are you _alright_ with this or would you prefer a brand new issue to address including the one-time 2FA pass?

**Comment by parasharrajat** [comment]
@mallenexpensify We can process the payment for this at it is pay day before proceeding to follow up issue.

**Comment by mallenexpensify** [comment]
@parasharrajat This issue is a bit confusing, I paid $500 for the initial fix, is more compensation due? If so, can you provide details?

**Comment by parasharrajat** [comment]
Yeah, Looks like this went through both holds (n6 and company offsite).

**Comment by mallenexpensify** [comment]
Thanks @parasharrajat It's def clear in this issue that n6-hold is applicable
![image](https://user-images.githubusercontent.com/22508304/146414368-36d527e3-23e5-44c6-9ff5-e22a1e848596.png)

For `Company offsite hold` (where I didn't properly assign the label to issues that were currently in-progress, it appears that the timeline for that is also applicable, as your PR wasn't reviewed until we lifted it. 

![image](https://user-images.githubusercontent.com/22508304/146414493-94a62498-1c4e-4959-8684-2802b660f0e7.png)

$500 Bonus added to the original job that I already paid $500 on.  

Closing
